File: py/doubly_linked_list/doubly_linked_list.py
class DoublyLinkedNode:
    def __init__(self, value):
        self.value = value
        self.next = None
        self.previous = None
    
class DoublyLinkedList:
    def __init__(self):
        self.head = None
        self.tail = None
        self.length = 0
    
    def prepend(self, value):
        list_node = DoublyLinkedNode(value)
        if not self.head:
            self.head = list_node
            self.tail = list_node
        else:
            self.head.previous = list_node
            list_node.next = self.head
        self.head = list_node
        self.length += 1
        return True
    
    def append(self, value):
        list_node = DoublyLinkedNode(value)
        if not self.head:
            self.head = list_node
            self.tail = list_node
            self.length += 1
        if self.tail:
            self.tail.next = list_node
            list_node.previous = self.tail
            self.tail = list_node
            self.length +=1
        return True
    
    def delete(self, value):
        if not self.head: return None
        current = self.head

        if self.head.value ==value and self.tail.value == value: 
            self.head.next = null
            self.tail.previous = null
        
        elif self.head.value == value:
            self.head = self.head.next
            self.head.previous = None
        
        elif self.tail.value == value:
            self.tail = self.tail.previous
            self.tail.next = None
        
        else:
            while current.value != value and current:
                current = current.next
            
            if current: 
                current.previous.next = current.next
                current.next.previous = current.previous
            else: return None
        self.length -=1
        return True

    def printList(self):
        if not self.head: return False
        current = self.head
        result = ""

        while current is not None:
            result += str(current.value) + "<=>"
            current = current.next
        
        print(result)
        return None

    def insert_at_position(self, position, data):
        counter =0
        current = self.head

        if not self.head:
            self.prepend(data)

        if position == 0:
            self.prepend(data)
        elif position == self.length -1:
            self.append(data)
        elif position > 0 and position < self.length -1:
            
            while counter != position:
                counter += 1
                current = current.next

            list_node = DoublyLinkedNode(data)
            current.previous.next = list_node
            list_node.previous = current
            list_node.next = current
            current.previous = list_node
            self.length +=1
                
        else: return None
    
    # def reversePrintList(self):
    #     if not self.head: return None
    #     current, next_node, prev_node = self.head, 0, 0
        
    #     while current is not None:
    #         prev_node = current.previous
    #         next_node = current.next
            
    #         current.next = prev_node
    #         current.previous = next_node
            
    #         prev_node = current
    #         current = next_node
            
    #     self.tail = self.head
    #     self.head = prev_node

    #     self.printList()
    #     return True
        

File: ts/src/doublyLinkedList/doublyLinkedList.ts
class DoubleLinkedListNode {
    public value: any
    public next: any
    public previous: any

    constructor(value: any, next: any = null, prev: any = null) {
        this.value = value
        this.next = next
        this.previous = prev
    }
}

class DoublyLinkedList {
    public head: DoubleLinkedListNode
    public length: number
    public tail: DoubleLinkedListNode

    constructor() {
        this.head = null
        this.tail = null
        this.length = 0
    }

    prepend(value: any): DoubleLinkedListNode {
        var listNode = new DoubleLinkedListNode(value)

        if (this.head) {
            this.head.previous = listNode
            listNode.next = this.head
        }

        this.head = listNode

        if (!this.tail) {
            this.tail = listNode
        }

        this.length += 1
        return listNode
    }

    append(value: any): DoubleLinkedListNode {
        var listNode = new DoubleLinkedListNode(value)

        if (!this.head) {
            this.head = listNode
            this.tail = listNode
            this.length += 1
        }

        if (this.tail) {
            this.tail.next = listNode
            listNode.previous = this.tail
            this.tail = listNode
            this.length += 1
        }
        return listNode
    }

    delete(value: any): boolean {
        if (!this.head) return null;
        var current = this.head

        if (this.head.value === value && this.tail.value === value) {
            this.head.next = null
            this.head.previous = null
            this.length -= 1
            return true
        } else if (this.head.value === value) {
            this.head = this.head.next
            this.head.previous = null
            this.length -= 1
            return true
        } else if (this.tail.value === value) {
            this.tail = this.tail.previous
            this.tail.next = null
            this.length -= 1
            return true
        } else {
            while (current.value != value && current != null) {
                current = current.next
            }

            if (current) {
                current.previous.next = current.next
                current.next.previous = current.previous
                this.length -= 1
                return true
            } else {
                return undefined
            }
        }
    }

    printList(): void {
        var current = this.head, result = ""
        if (!this.head) return undefined

        while (current != null) {
            result += current.value + "<=>"
            current = current.next
        }
        console.log(result)
        return null
    }

    reversePrintList(): void {
        if (!this.head) return undefined
        var current = this.head, nextNode, prevNode

        while (current) {
            prevNode = current.previous
            nextNode = current.next

            current.next = prevNode
            current.previous = nextNode

            prevNode = current
            current = nextNode
        }

        this.tail = this.head
        this.head = prevNode
        return null
    }
}

export default DoublyLinkedList